docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,527 @@
+* Reactor: A design pattern loosely defined as "something which listens on a channel and reacts to messages".
+
+These abstractions and related concepts are illustrated in the following diagram and described in detail below.
+
+![P2P Architecture Diagram](img/adr-062-architecture.svg)
+
+### Transports
+
+Transports are arbitrary mechanisms for exchanging raw bytes with a peer. For example, a gRPC transport would connect to a peer over TCP/IP and send data using the gRPC protocol, while an in-memory transport might communicate with a peer running in another goroutine using internal byte buffers. Note that transports don't have a notion of a `peer` as such - instead, they communicate with arbitrary endpoint addresses (e.g. IP address and port number), to decouple them from the rest of the P2P stack.
+
+Transports must satisfy the following requirements:
+
+* Be connection-oriented, and support both listening for inbound connections and making outbound connections using endpoint addresses.
+
+* Support multiple logical IO streams within a single connection, to take full advantage of protocols with native stream support. For example, QUIC supports multiple independent streams, while HTTP/2 and MConn multiplex logical streams onto a single TCP connection.
+
+* Provide the public key of the peer, and possibly encrypt or sign the traffic as appropriate. This should be compared with known data (e.g. the peer ID) to authenticate the peer and avoid man-in-the-middle attacks.
+
+The initial transport implementation will be a port of the current MConn protocol currently used by Tendermint, and should be backwards-compatible at the wire level. This will be followed by an in-memory transport for testing, and a QUIC transport that may eventually replace MConn.
+
+The `Transport` interface is:
+
+```go
+// Transport is an arbitrary mechanism for exchanging bytes with a peer.
+type Transport interface {
+    // Accept waits for the next inbound connection in a listening endpoint.
+    Accept(context.Context) (Connection, error)
+
+    // Dial creates an outbound connection to an endpoint.
+    Dial(context.Context, Endpoint) (Connection, error)
+
+    // Endpoints lists endpoints the transport is listening on. Any endpoint IP
+    // addresses do not need to be normalized in any way (e.g. 0.0.0.0 is
+    // valid), as they should be preprocessed before being advertised.
+    Endpoints() []Endpoint
+}
+```
+
+How the transport configures listening is transport-dependent, and not covered by the interface. This typically happens during transport construction.
+
+#### Endpoints
+
+`Endpoint` represents a transport endpoint (e.g. an IP address and port). A connection always has two endpoints: one at the local node and one at the remote peer. An outbound connection to a remote endpoint is made via a `Dial()` call, and inbound connections to listening endpoints are returned via `Accept()`.
```

Comment:
Should this be extended to support proxy endpoints?

Reply:
In what way? Like forward proxies?

Reply:
yea, is this covered with uPnP?

Reply:
No, UPnP is for NAT gateway autoconfiguration. I'm not really sure I understand the use-case for a forward proxy, I thought people stopped using those in the 90s?

Reply:
I guess you maybe mean some way to advertise the external endpoint of something that's behind e.g. a NAT gateway? That's somewhat orthogonal, since there will be a separate endpoint address for the internal and external address. This will need to be handled in the system that does address advertisement and peer exchange - I've tried to keep it in mind here, but that system needs to be spec'ed out separately. Basically, it will collect all listening endpoints from all local addresses, and then based on those try to determine which ones are internal and which ones are external, and try to detect or translate the internal ones into external addresses (via e.g. UPnP, configuration options, or other mechanisms).